src/infrastructure/backends/githubcopilot_test.rs

```diff
@@ -0,0 +1,405 @@
+        insta::assert_toml_snapshot!(third_recv.context);
+    });
+
+    return Ok(());
+}
+
+#[tokio::test]
+async fn it_gets_completions_token_expired() -> Result<()> {
+    let new_token = "token_expired".to_string();
+    let current_expires_at: u64 = 1577907680; // Expired in 2020
+    let MockGithubCopilot {
+        comp_server,
+        comp_mock,
+        auth_server,
+        auth_mock,
+        backend,
+    } = setup_get_completion("oauth-12345".to_string(), new_token.clone())?;
+
+    let (tx, mut rx) = mpsc::unbounded_channel::<Event>();
+    let prompt = BackendPrompt {
+        text: "Say hi to the world".to_string(),
+        backend_context: serde_json::to_string(&vec![
+            MessageRequest {
+                role: "system".to_string(),
+                content: "How may I help you?".to_string(),
+            },
+            MessageRequest {
+                role: "__token".to_string(),
+                content: serde_json::to_string(&CopilotTokenResponse {
+                    token: "abc".to_string(),       // Old token
+                    expires_at: current_expires_at, // Expired in 2020
+                    chat_enabled: true,
+                })?,
+            },
+        ])?,
+    };
+    backend.get_completion(prompt, &tx).await?;
+
+    comp_mock.assert();
+    auth_mock.assert();
+
+    let first_recv = to_res(rx.recv().await)?;
+    let second_recv = to_res(rx.recv().await)?;
+    let third_recv = to_res(rx.recv().await)?;
+
+    assert_eq!(first_recv.author, Author::Model);
+    assert_eq!(first_recv.text, "Hello ".to_string());
+    assert!(!first_recv.done);
+    assert_eq!(first_recv.context, None);
+
+    assert_eq!(second_recv.author, Author::Model);
+    assert_eq!(second_recv.text, "World".to_string());
+    assert!(!second_recv.done);
+    assert_eq!(second_recv.context, None);
+
+    assert_eq!(third_recv.author, Author::Model);
+    assert!(third_recv.text.is_empty());
+    assert!(third_recv.done);
+
+    let context = third_recv.clone().context.expect("Context must be defined");
+
+    let context_messages: Vec<MessageRequest> = serde_json::from_str(&context)?;
+    let index = context_messages
+        .iter()
+        .position(|value| return value.role == *"__token")
+        .expect("Token must be set");
+
+    let msg = context_messages.get(index).expect("Token must be set");
+    let content = serde_json::from_str::<CopilotTokenResponse>(&msg.content)?;
+    assert_eq!(content.token, new_token);
+    assert!(content.expires_at > current_expires_at);
+
+    insta_snapshot(|| {
+        insta::assert_toml_snapshot!(third_recv.context);
+    });
+
+    return Ok(());
+}
+
+#[tokio::test]
+async fn it_gets_completions() -> Result<()> {
+    let token = "token_ok".to_string();
+    let MockGithubCopilot {
+        comp_server,
+        comp_mock,
+        auth_server,
+        auth_mock,
+        backend,
+    } = setup_get_completion("oauth-12345".to_string(), token.clone())?;
```

Comment:
Here I am using a single DRY setup for the oauth and token cases. 

I am returning the unused comp_server and auth_server, to extend their corresponding lifetimes into each of the tests. If I don't do this, the servers become unavailable and the tests fail. The  linter removes this values as they are unused so it will naturally fail on CI. 

Also, creating the servers and mocks 4 times looks unnecessary, although I do rely on the assert/expect(n) for the tests. How would you recommend handling this?

Reply:
> The linter removes this values as they are unused so it will naturally fail on CI.

Prefixing variable names with underscores will prevent them from being removed.

>Also, creating the servers and mocks 4 times looks unnecessary,

I agree, but in Rust when using nextest you're kinda stuck with it. There's no setup/teardown steps, and each test is ran in it's own process.